### PR TITLE
feat: implement dynamic gas limit analysis in approveProposal function

### PR DESCRIPTION
**Pull request by jyotirmoydotdev**
# Gas Estimation Strategy

This pull request introduces enhancements to the **`approveProposal`** function in the smart contract to include a gas limit analysis strategy. The goal is to ensure that the transaction executes within reasonable computational constraints by dynamically estimating gas limits based on the length of the proposal target array.

## Conditional Check

```solidity
if ((proposals[_proposalId].approvals + 1) == 3)
```

This checks if the number of approvals for the specified proposal (**`_proposalId`**) plus 1 equals 3. In other words, it checks if the total number of approvals after the current one reaches the quorum, which is set to 3.

## Gas Estimation

```solidity
// Gas estimation for the first proposal
uint256 baseGas = 110539;

// Average gas difference for each additional proposal
uint256 averageGasDifference = 31560;

// Safety margin percentage (e.g., 10%)
uint256 safetyMarginPercentage = 10;
```

These variables are used for estimating the gas limit. **`baseGas`** represents the estimated gas for the first transaction, **`averageGasDifference`** is the average increase in gas for each additional transaction, and **`safetyMarginPercentage`** is the percentage of safety margin to be added to the gas limit.

## Ensure Proposal Array is Not Empty

```solidity
// Ensure proposal array (target) is not empty
require(proposals[_proposalId].target.length > 0, "Proposal array is empty");
```

This line ensures that the proposal array (**`target`**) associated with the specified proposal is not empty. If the array is empty, the function will revert with the error message "Proposal array is empty."

## Calculate Expected Gas Limit

```solidity
// Calculate the expected gas limit for the entire transaction
uint256 expectedGasLimit = baseGas + (proposals[_proposalId].target.length - 1) * averageGasDifference;
```

This calculates the expected gas limit for the entire transaction. It considers the length of the proposal target array (**`proposals[_proposalId].target`**) and multiplies the average gas difference by the number of additional proposals (length - 1).

## Add Safety Margin

```solidity
// Add a safety margin to the expected gas limit
uint256 safetyMargin = (expectedGasLimit * safetyMarginPercentage) / 100;
uint256 totalGasLimit = expectedGasLimit + safetyMargin;
```

The code calculates a safety margin based on the percentage specified in **`safetyMarginPercentage`** and adds it to the expected gas limit. The result is stored in **`totalGasLimit`**.

## Check Gas Limit

```solidity
// Ensure the actual gas used matches the expected gas limit with safety margin
require(gasleft() >= totalGasLimit, "Insufficient gas limit");
```

Finally, this line ensures that the actual gas remaining (**`gasleft()`**) during the execution of the function is greater than or equal to the calculated total gas limit. If the gas limit is not sufficient, the function will revert with the error message "Insufficient gas limit."

In summary, this code block performs gas estimation and checks if the remaining gas is sufficient for the transaction to ensure it doesn't run out of gas during execution. It adds a safety margin to the gas limit to account for potential variations in gas consumption.

### CODE REVIEW

File: src/Chamber.sol [hunk 1]
```diff
@@ -106,6 +106,30 @@ contract Chamber is IChamber, Common {
 
         if (!onVoterList) revert invalidApproval("TokenId not on voter list");
 
+        if ((proposals[_proposalId].approvals+1)==3){
+            // Gas estimation for the first proposal
+            uint256 baseGas = 110539;
```
Comment (xhad) [inline src/Chamber.sol:111]: Where are you getting this value from? How can you be certain this value is going to be valid for a wide range of possible transactions?

File: src/Chamber.sol [hunk 2]
```diff
@@ -106,6 +106,30 @@ contract Chamber is IChamber, Common {
...
+
+            // Average gas difference for each additional proposal
+            uint256 averageGasDifference = 31560;
```
Comment (xhad) [inline src/Chamber.sol:114]: Where did you get this value from, and how can you be sure this is valid?

**Comment by jyotirmoydotdev** [comment]
I obtained these values (baseGas and averageGasDifference) from the test. Initially, I set the length of the proposals(_proposalId).target array to 1, containing only one target, one value, and one data. After conducting the test and calculating the gas limit using gasleft(), I obtained my first result.

```solidity
//....

uint256 gasBefore = gasLeft();
chamber.approveProposal(1, 3, getSignature(1, 3, 1));
uint256 gasAfter = gasLeft();
uint256 gasUsage = gasBefore - gasAfter;
//....
```

| Proposal[ID].target | Approve 1 | Approve 2 | Approve 3 |
| --- | --- | --- | --- |
| Target Length = 1 | 83570 | 63670 | 110539 |

This is the result of the first test, where the "Approve 1" column denotes how much gas was used when the first leader approved the proposal, "Approve 2" for the second approval, and "Approve 3" for the third one. You can see that the third approval cost a high gas amount because when the third leader approved the proposal, they also executed it.

Since the proposal cannot have 0 transactions and must have at least 1 transaction in the proposals target array, I consider 110539 as the `baseGas`.

So, after obtaining the baseGas, I increased the target array size to 2 (also value and data), ran the test again, and got the following result. The result indicates that when the third leader approved the proposal, the gas cost increased by 27,290 and, for safety, was rounded off to 27300.

| Proposal[ID].target | Approve 1 | Approve 2 | Approve 3 |
| --- | --- | --- | --- |
| Target Length = 1 | 83570 | 63670 | 110539 |
| Target Length = 2 | 85113 | 65213 | 137829 |

After increasing the target length to 3 and 4, the results were as follows:

| Proposal[ID].target | Approve 1 | Approve 2 | Approve 3 |
| --- | --- | --- | --- |
| Target Length = 1 | 83570 | 63670 | 110539 |
| Target Length = 2 | 85113 | 65213 | 137829 |
| Target Length = 3 | 85611 | 65711 | 172769 |
| Target Length = 4 | 86111 | 66211 | 205209 |

To find how much gas increased with each added transaction, we can sum up the individual differences for "Approve 3" and then divide by the number of differences, which is 3.

_Average Difference_ = ((137829 - 110539) + (172769 - 137829) + (205209 - 172769)) / 3

_Average Difference_ = (27300 + 34940 + 32440) / 3

_Average Difference_ = 94680 / 3

_Average Difference_ = 31560

So, the `averageGasDifference` in the sequence is 31,560.

This way, we can dynamically calculate the gas cost based on the target array length, which would be `baseGas + (proposals[_proposalId].target.length - 1) * averageGasDifference`.

File: src/Chamber.sol (hunk 2 above)
Comment (jyotirmoydotdev) [reply src/Chamber.sol:114]: Please read this 
https://github.com/loreum-org/chamber/pull/47#issuecomment-1905242615

File: src/Chamber.sol (hunk 1 above)
Comment (jyotirmoydotdev) [reply src/Chamber.sol:111]: Please read this 
https://github.com/loreum-org/chamber/pull/47#issuecomment-1905242615

**Comment by jyotirmoydotdev** [comment]
I also implemented a check so that this gas requirement only applies to the third leader.

```solidity
if ((proposals[_proposalId].approvals + 1) == 3) {
```
